### PR TITLE
Streaming writes

### PR DESCRIPTION
**Pull request by davidvgalbraith**
Building on https://github.com/juttle/juttle-s3-adapter/pull/1, use the serializers to support CSV and JSONL as well as JSON. 

Fixes https://github.com/juttle/juttle-s3-adapter/issues/2 and https://github.com/juttle/juttle-s3-adapter/issues/3.

@juttle/developers

### CODE REVIEW

**Comment by davidvgalbraith** [comment]
@VladVega

File: README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+# Juttle S3 Adapter
+
+[![Build Status](https://travis-ci.org/juttle/juttle-s3-adapter.svg)](https://travis-ci.org/juttle/s3-adapter)
```
Comment (VladVega) [inline README.md]: `https://travis-ci.org/juttle/juttle-s3-adapter.svg?branch=master`

File: lib/s3-client.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        this.client = Promise.promisifyAll(new AWS.S3({
+            accessKeyId: accessKeyId,
+            secretAccessKey: secretAccessKey,
+            region: region
+        }));
+    }
+
+    set_or_throw(option, value) {
+        if (!value) {
+            throw new Error('S3Client missing required option ' + option);
```
Comment (VladVega) [inline lib/s3-client.js]: Consider using reusable error for this? https://github.com/juttle/juttle/blob/master/lib/adapters/file/read.js#L18

File: lib/s3-client.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        }
+
+        this[option] = value;
+    }
+
+    read(limit) {
```
Comment (VladVega) [inline lib/s3-client.js]: Won't `limit` here actually be `from` since the interface is `read(from, to, limit, state)`?

File: lib/s3-client.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        return new Promise((resolve, reject) => {
+            var points = [];
+            var request = this.client.getObject({
+                Bucket: this.bucket,
+                Key: this.key,
+            });
+
+            request.on('httpHeaders', (statusCode, headers, resp) => {
+                if (statusCode >= 300) {
+                    throw (new Error("http status code " + statusCode));
```
Comment (VladVega) [inline lib/s3-client.js]: Is there is an error message that you can send up with the code? It's probably in the http body.

File: package.json [hunk 5]
```diff
@@ -14,5 +14,12 @@
...
   ],
   "author": "davidvgalbraith",
-  "license": "ISC"
+  "license": "ISC",
+  "dependencies": {
+    "aws-sdk": "^2.2.32",
+    "bluebird": "^3.1.5"
+  },
+  "devDependencies": {
+    "chai": "^3.4.1"
```
Comment (VladVega) [inline package.json]: peer dependency on juttle?

File: lib/s3-client.js (hunk 3 above)
Comment (davidvgalbraith) [inline lib/s3-client.js]: This is `s3-client.js`. The from/to/limit/state arguments are handled in `read.js` at https://github.com/juttle/juttle-s3-adapter/pull/5/files#diff-77960d7f37d725c98368151aa4bc18f7R22.

Comment (VladVega) [inline lib/s3-client.js]: Oh my bad, got it.

File: lib/s3-client.js (hunk 4 above)
Comment (davidvgalbraith) [inline lib/s3-client.js]: Nope! This is the httpHeaders event, which fires once we get just the headers and doesn't have any body.

File: lib/s3-client.js (hunk 2 above)
Comment (davidvgalbraith) [inline lib/s3-client.js]: Let's do required option validation in the base class like https://github.com/juttle/juttle/pull/317.

**Pull request merged**